File: ruby_practice_problems/04_search_engine.rb/snapshot_array.rb
class SnapshotArray
  attr_accessor :arr, :snap_hash
  def initialize(length)
    @arr = [0] * length
    @snap_id = 0
    @snap_hash = {}
  end

  def set_value(idx, val)
    @arr[idx] = val
  end

  def snapshot()
    @snap_hash[@snap_id] = @arr.map(&:itself)
    @snap_id += 1
    return @snap_id - 1
  end

  def get_value(idx, snapid)
    if @snap_hash[snapid] == nil || idx > @snap_hash[snapid].length()
      return nil
    else
      return @snap_hash[snapid][idx]
    end
  end
end

File: python_practice_problems/24_custom_data_structures/snapshot_array.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed May  3 18:37:04 2023

@author: johnmorgan
"""

class SnapshotArray:
    def __init__(self, length):
        self.arr = [0] * length
        self.snap_dict = {}
        self.snap_counter = 0
        
    def set_value(self, idx, val):
        self.arr[idx] = val
        
    def snapshot(self):
        snap_arr = [0] * len(self.arr)
        for i, elem in enumerate(self.arr):
            snap_arr[i] = elem
        self.snap_dict[self.snap_counter] = snap_arr 
        self.snap_counter += 1
        return self.snap_counter - 1
    
    def get_value(self, idx, snap_id):
        return self.snap_dict[snap_id][idx]
    
    